insights/specs/datasources/pcp.py

```diff
@@ -0,0 +1,82 @@
+"""
+Custom datasource related PCP (Performance Co-Pilot)
+"""
+import logging
+import datetime
+import os
+
+from insights.core.dr import SkipComponent
+from insights.core.context import HostContext
+from insights.core.plugins import datasource
+from insights.parsers.ros_config import RosConfig
+from insights.combiners.ps import Ps
+from insights.combiners.services import Services
+
+logger = logging.getLogger(__name__)
+
+
+@datasource(Services, HostContext)
+def pcp_enabled(broker):
+    """
+    Returns:
+        bool: True if pmproxy service is on in services
+
+    Raises:
+        SkipComponent: When pmproxy service is not enabled
+    """
+    if not broker[Services].is_on("pmproxy"):
+        raise SkipComponent("pmproxy not enabled")
+    return True
+
+
+@datasource(Ps, HostContext)
+def pmlog_summary_file(broker):
+    """
+    Determines the name for the pmlogger file and checks for its existence
+
+    Returns the name of the latest pmlogger summary file if a running ``pmlogger``
+    process is detected on the system.
+
+    Returns:
+        str: Full path to the latest pmlogger file
+
+    Raises:
+        SkipComponent: Raises this exception when the command is not present or
+            the file is not present
+    """
+    ps = broker[Ps]
+    if ps.search(COMMAND_NAME__contains='pmlogger'):
+        pcp_log_date = (datetime.date.today() - datetime.timedelta(days=1)).strftime("%Y%m%d")
+        file = "/var/log/pcp/pmlogger/ros/%s.index" % (pcp_log_date)
+        try:
+            if os.path.exists(file) and os.path.isfile(file):
+                return file
+        except Exception as e:
+            SkipComponent("Failed to check for pmlogger file existence: {0}".format(str(e)))
+
+    raise SkipComponent
+
+
+@datasource(RosConfig, HostContext)
+def pmlog_summary_metrics(broker):
+    """
+    Returns the metrics specified as "mandatory on" in the `config.ros` file
+    for the pmlogger summary file.
+
+    Returns:
+        str: Full path to the latest pmlogger file
+
+    Raises:
+        SkipComponent: When no metrics get from the RosConfig
+    """
+    ros = broker[RosConfig]
+    metrics = set()
+
+    for spec in ros.specs:
```

Comment:
One query - do we need to handle `null` condition here when ROS config file absent on targeted system?

Reply:
@kgaikwad, Thanks for reviewing.

Since the `RosConfig` is a `Required` component of the datasource, when the `Required` component is not triggered, this datasoruce will cannot be trigger too.

See: https://insights-core.readthedocs.io/en/latest/api_index.html#insights.core.dr.ComponentType

TLDR;, see the following test:

```python
# sudo mv /var/lib/pcp/config/pmlogger/config.ros /var/lib/pcp/config/pmlogger/config.ros.bak
# ls /var/lib/pcp/config/pmlogger/config.ros
ls: cannot access '/var/lib/pcp/config/pmlogger/config.ros': No such file or directory
# insights-inspect insights.specs.Specs.pmlog_summary
Missing Dependencies:
    At Least One Of:
        insights.specs.default.DefaultSpecs.pmlog_summary
# ^^ The spec will not get fired when missing the `config.ros`
# sudo mv /var/lib/pcp/config/pmlogger/config.ros.bak /var/lib/pcp/config/pmlogger/config.ros
# ^^ Get the `config.ros` file back
# insights-inspect insights.specs.Specs.pmlog_summary

IPython Console Usage Info:

Enter 'pmlog_summary.' and tab to get a list of properties
Example:
In [1]: pmlog_summary.<property_name>
Out[1]: <property value>

To exit ipython enter 'exit' and hit enter or use 'CTL D'

Starting IPython Interpreter Now

In [1]: !ls /var/lib/pcp/config/pmlogger/config.ros
/var/lib/pcp/config/pmlogger/config.ros

In [2]: pmlog_summary.content
Out[2]:
['disk.all.total  0.401 count / sec',
 'kernel.all.cpu.idle  3.973 none',
 'kernel.all.cpu.nice  0.003 none',
 'kernel.all.cpu.steal  0.000 none',
 'kernel.all.cpu.sys  0.005 none',
 'kernel.all.cpu.user  0.012 none',
 'kernel.all.cpu.wait.total  0.000 none',
 'mem.physmem  7835748.000 Kbyte',
 'mem.util.bufmem  5236.000 Kbyte',
 'mem.util.cached  2572533.667 Kbyte',
 'mem.util.free  690659.853 Kbyte',
 'mem.util.used  7145088.147 Kbyte']
```

Reply:
Thank you @xiangce for the details.

Reply:
That's what I missed for datasource, and according to Bob's comment, I used `try/except` for this check. Sorry for the incomplete explanation.